Future[T] and Promise[T] as Controller ReturnTypes

domdorn opened the pull request:
This PR adds support for `Future[T]` and `Promise[T]` as return types in resources, e.g. 
```scala

@Path("/")
class GreetingResource() 
    @GET
    @Path("/greet/promise")
    def promiseGreeting(): Promise[String] =
      Promise.successful("Hello from the promise")
    end promiseGreeting
end GreetingResource
```

I also added tests in the integration module and converted them to be in Scala

domdorn committed:
Support for using `Future[T]` or `Promise[T]` as response type in Rest Resources

domdorn commented:
I've reworked this PR so that everything that resolves around using `Future[T]` or `Promise[T]` is in it own sub-extension. In another PR I'll add support for `ZIO[R,E,A]` in the same manner.

gsmet commented on README.md:
```diff
@@ -61,6 +61,7 @@ VERSIONS = [
 
 dependencies {
     implementation "io.quarkiverse.scala:quarkus-scala3:${VERSIONS.QUARKUS_SCALA3}"
+    implementation "io.quarkiverse.scala:quarkus-scala3-deployment:${VERSIONS.QUARKUS_SCALA3}"
```
I don't think this is right. You shouldn't have to add the `-deployment` dependencies.

domdorn replied:
done

gsmet commented on README.md:
```diff
@@ -133,7 +134,12 @@ In your `pom.xml` file, add:
-    <version>0.0.1<version>
+    <version>1.0.0<version>
+</dependency>
+<dependency>
+  <groupId>io.quarkiverse.scala</groupId>
+  <artifactId>quarkus-scala3-deployment</artifactId>
+  <version>1.0.0</version>
```
No, you don't need to do that. Why do you think it's needed?

domdorn replied:
removed it

gsmet commented on README.md:
````diff
@@ -440,6 +446,54 @@ def mkRoutes(router: Router) =
 ```
 
+
+### `Future[T]` and `Promise[T]` support in Rest-Endpoints
````
```suggestion
### `Future[T]` and `Promise[T]` support in REST endpoints
```

gsmet commented on README.md:
```diff
@@ -440,6 +446,54 @@ def mkRoutes(router: Router) =
+
+### `Future[T]` and `Promise[T]` support in Rest-Endpoints
+
+This extension allows you to return `Future[T]` and `Promise[T]` from your rest-endpoints.
```
```suggestion
This extension allows you to return `Future[T]` and `Promise[T]` from your REST endpoints.
```

gsmet commented on README.md:
```diff
@@ -440,6 +446,54 @@ def mkRoutes(router: Router) =
+      <artifactId>quarkus-scala3</artifactId>
+      <version>${project.version}</version>
+    </dependency>
+    <dependency>
+      <groupId>io.quarkiverse.scala</groupId>
+      <artifactId>quarkus-scala3-deployment</artifactId>
+      <version>${project.version}</version>
+    </dependency>
```
This will need to go too.

gsmet commented on README.md:
```diff
@@ -452,6 +506,7 @@ Thanks goes to these wonderful people ([emoji key](https://allcontributors.org/d
     <tr>
       <td align="center" valign="top" width="14.28%"><a href="https://github.com/GavinRay97"><img src="https://avatars.githubusercontent.com/u/26604994?v=4?s=100" width="100px;" alt="Gavin Ray"/><br /><sub><b>Gavin Ray</b></sub></a><br /><a href="https://github.com/quarkiverse/quarkus-scala3/commits?author=GavinRay97" title="Code">💻</a> <a href="#maintenance-GavinRay97" title="Maintenance">🚧</a>< ...
       <td align="center" valign="top" width="14.28%"><a href="https://lesincroyableslivres.fr/"><img src="https://avatars.githubusercontent.com/u/1279749?v=4?s=100" width="100px;" alt="Guillaume Smet"/><br /><sub><b>Guillaume Smet</b></sub></a><br /><a href="https://github.com/quarkiverse/quarkus-scala3/commits?author=gsmet" title="Code">💻</a></td>
+      <td align="center" valign="top" width="14.28%"><a href="https://github.com/domdorn"><img src="https://avatars.githubusercontent.com/u/100349?v=4?s=100" width="100px;" alt="Dominik Dorn"/><br /><sub><b>Dominik Dorn</b></sub></a><br /><a href="https://github.com/quarkiverse/quarkus-scala3/commits?author=domdorn" title="Code">🚀🍺💻</a></td>
```
I would revert that, We will ask the allcontributors bot to add you.

gsmet commented on README.md:
```diff
@@ -462,3 +517,7 @@ Thanks goes to these wonderful people ([emoji key](https://allcontributors.org/d
+
+## TODOs
+ - correctly generate OpenAPI Spec for methods returning `Future[T]` or `Promise[T]`, e.g. similar to [Quarkus #8499](https://github.com/quarkusio/quarkus/issues/8499)
+ - Quarkus-Arc has special handling for `CompletionStage[T]`, maybe we should add similar handling for `Future[T]` and `Promise[T]`, see [ActiveRequestContextInterceptor](https://github.com/quarkusio/quarkus/blob/24d3e5262d20fdaa8c056d59f012f8c7b5b1c5c8/independent-projects/arc/runtime/src/main/java/io/quarkus/arc/impl/ActivateRequestContextInterceptor.java) ?
```
```suggestion
 - ArC (Quarkus' CDI implementation) has special handling for `CompletionStage[T]`, maybe we should add similar handling for `Future[T]` and `Promise[T]`, see [ActiveRequestContextInterceptor](https://github.com/quarkusio/quarkus/blob/24d3e5262d20fdaa8c056d59f012f8c7b5b1c5c8/independent-projects/arc/runtime/src/main/java/io/quarkus/arc/impl/ActivateRequestContextInterceptor.java) ?
```

gsmet commented on futures/deployment/src/main/scala/io/quarkiverse/scala/scala3/futures/deployment/Scala3FutureResponseHandler.java:
```diff
@@ -0,0 +1,34 @@
+import org.jboss.resteasy.reactive.server.core.ResteasyReactiveRequestContext;
+import org.jboss.resteasy.reactive.server.spi.ServerRestHandler;
+
+public class Scala3FutureResponseHandler implements ServerRestHandler {
```
I think this is probably why you had to add the -deployment dependencies in your build files: this runtime class should be in a `runtime` module.

The `-deployment` module is only around at build time, anything that is used at runtime has to be in the runtime module.

gsmet commented on futures/runtime/src/main/resources/META-INF/quarkus-extension.yaml:
```diff
@@ -0,0 +1,11 @@
+name: Scala 3 Futures Support
+description: Support for using `Future[T]` and `Promise[T]` in REST Endpoints.
```
I'm not sure we are using Markdown in the tooling everywhere, so I would remove the backticks.

domdorn replied:
done

gsmet commented on futures/deployment/src/main/scala/io/quarkiverse/scala/scala3/futures/deployment/Scala3FutureResponseHandler.java:
```diff
@@ -0,0 +1,34 @@
+package io.quarkiverse.scala.scala3.deployment;
```
The package name doesn't look good.

gsmet commented on futures/deployment/src/main/scala/io/quarkiverse/scala/scala3/futures/deployment/Scala3FutureReturnTypeMethodScanner.java:
```diff
@@ -0,0 +1,50 @@
+package io.quarkiverse.scala.scala3.deployment;
```
Same here about the package name.

gsmet reviewed (changes requested):
Thanks for the PRs, it's nice to see some activity in this project.

I added some suggestions, some of them minor, some of them not so much :).

gsmet commented:
BTW, I suggest we get this one in shape and merged, before applying the same changes to the ZIO PR.

domdorn committed:
Update README.md

Co-authored-by: Guillaume Smet <guillaume.smet@gmail.com>

domdorn committed:
Update README.md

Co-authored-by: Guillaume Smet <guillaume.smet@gmail.com>

domdorn committed:
Update README.md

Co-authored-by: Guillaume Smet <guillaume.smet@gmail.com>

gsmet commented:
Disclaimer: I have no knowledge of Scala, so please don't hesitate to push back if what I say doesn't make sense :).

BTW, I think you could put everything in the same module.

The build item machinery should be clever enough to not trigger the loading of the classes you added if the Quarkus REST extension is not around.
My guess is that you could simplify the integration by having the SPI dependencies around but `quarkus-rest` would be an optional dependency.

I think that would have my preference if you can make it work as the experience will be better for the users.

domdorn committed:
futures: PR Feedback incorporated

domdorn commented:
@gsmet I've updated the PR, please have a look. I noticed that I don't get the stack trace in the native executable in case of an error, so I removed the check on the body.
I had everything in the same extension but when adding the zio one I figured it would be cleaner to have it in its own extension, especially if one day we'll add other effect types (cats/monix/whatever) as well..

domdorn committed:
PR feedback integrated

carlosedp commented:
Hey folks, any news on this?

domdorn commented:
I think I've applied everything that was requested, but maybe I missed something. 
Anyways, I'm going to be on holidays the next 3 weeks without spare time to work on this. 
From my point of view, the potentially remaining issues are just about moving code to the right places. 
If anyone wants to fork/take over the PR, please do so!

carlosedp commented:
Hi all... any news from this? It would be great to have it in the extension.
